Added ':' in environment variable name valid characters list

vithati opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR changes allows ':' in environment variable name for the containers. This is bug fix of issue
https://github.com/kubernetes/kubectl/issues/47 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubectl/issues/47

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
Allowed ':' in environment variable names
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Relaxed restrictions on environment variable name characters. A valid environment variable name should not start with a digit. Older version of Kubelet ignores the environment variable names having other than [-._a-zA-Z0-9] characters and using these variables names will prevent apiserver rollback to older versions.
```

vithati commented:
/sig cli
/area kubectl
/assign @apelisse

vithati commented:
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please email the CNCF helpdesk: [helpdesk@rt.linuxfoundation.org](mailto:helpdesk@rt.linuxfoundation.org)

I have signed CLA.

apelisse commented:
/ok-to-test

You need to sign the CLA if you want your code to be merged.

apelisse commented:
I think this is relaxing (rather than tightening) the behavior, so it shouldn't be a compatibility problem.
I also think it's POSIX compliant (see [here](https://github.com/kubernetes/kubectl/issues/47#issuecomment-321596303)).

We've had some problems using colons in some arguments in the past, can you maybe add a test in [test/cmd](https://github.com/kubernetes/kubernetes/tree/ae2efc752afb1f681334343a0dcd7cf66ea5e39e/test/cmd)?

Also, I suspect we might want to relax the regex even more, since I suspect this won't be enough to fix the issue you're trying to fix (the example does include a whitespace).

jennybuckley commented:
/cc @kubernetes/api-reviewers

lavalamp commented:
Rollback is unsafe as soon as someone makes a env named with a ":".

lavalamp commented:
Yes, this is an API change.

On Thu, Oct 4, 2018 at 2:25 PM Antoine Pelisse <notifications@github.com>
wrote:

> Oh, OK no, this is not just for kubectl, this is changing the apimachinery
> check.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/69415#issuecomment-427174680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngls3vpaeDFFXWNFDKQNzCFMMxyvRGks5uhnzbgaJpZM4XHqxe>
> .
>

apelisse commented:
> Rollback is unsafe as soon as someone makes a env named with a ":".

is this something we can't accept, or a release-note might be OK?

thockin commented:
We have relaxed this in the past (it was once very strict).  This is a POSIX compatible change though.

This is strictly rollback-unsafe, but I'd say it's relatively low-risk.  How do we want to manage these?  Do we say no to this forever?  Do we replace `Env` with `POSIXEnv`?

If we do this, we should also update the comments in pkg/apis/core/types.go and staging/src/k8s.io/api/core/v1/types.go

apelisse commented:
> This is strictly rollback-unsafe, but I'd say it's relatively low-risk.

I'd go with this, strongly suggest in the release note to delay using this variables (which people can't use today anyway) before they are confident they won't have to fallback their apiserver.

lavalamp commented:
That works for me, fix the release note & I'll approve.

(also looks like CLA needs signing)

apelisse commented:
Things that need to be fixed on this PR:
- [ ] Sign the CLA
- [ ] Update the release note to indicate that starting to use these variable names will prevent further roll-backs
- [ ] Update documentation, since keys don't have to be `C_IDENTIFIER`s anymore:
  - https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L1655
  - https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L1932
  - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1748
  - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1831
  - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2065

vithati commented:
I signed it

vithati commented:
> Things that need to be fixed on this PR:
> 
> * [ ]  Sign the CLA
> * [ ]  Update the release note to indicate that starting to use these variable names will prevent further roll-backs
> * [ ]  Update documentation, since keys don't have to be `C_IDENTIFIER`s anymore:
>   
>   * https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L1655
>   * https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L1932
>   * https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1748
>   * https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1831
>   * https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2065

@apelisse,  I have addressed above comments.

apelisse commented:
I think you should relax the regex even more, allowing more characters as long as it is POSIX compatible. Improve the documentation by shortly describing what's permitted: `Name of the environment variable. All characters are allowed, MUST NOT start with a number.`. I would also emphasize the risk of using new names in the release note.

liggitt commented:
> I think you should relax the regex even more, allowing more characters as long as it is POSIX compatible

discussion at https://github.com/kubernetes/kubernetes/pull/48986#issuecomment-315579043 seems relevant

apelisse commented:
> I'd probably prefer to have a set of concrete additions, each backed by a concrete use case.

I'm not sure I agree with that statement. We shouldn't decide what variables people can use, especially when POSIX reads: 
> Other characters may be permitted by an implementation; applications shall tolerate the presence of such names.

Obviously, we have to ensure that nodes/container runtime support these new variables.

vithati commented:
> I think you should relax the regex even more, allowing more characters as long as it is POSIX compatible. Improve the documentation by shortly describing what's permitted: `Name of the environment variable. All characters are allowed, MUST NOT start with a number.`

Do you mean, allow all character in environment key name except starting with number? if yes, https://github.com/kubernetes/kubernetes/pull/48986 restricted keys starting with "." and "..". If we allow all charcter, will it create issues?

apelisse commented:
@yujuhong @Random-Liu Do you know how well crazy variable names are supported on the CRI API and nodes? (I have no idea what I'm talking about, I'm repeating words Jordan used).

vithati commented:
/retest

vithati commented:
@yujuhong and @Random-Liu, If you want me to validate any CRI implementation with more relaxed set and can get me started on it, I can do that as well. 

@apelisse and @lavalamp, I am assuming that I should make the checks green on this PR and wait for @yujuhong and @Random-Liu to respond so that we can close on character set. Please do let me know in case there is something I can do to quickly close on this.

yujuhong commented:
> @yujuhong @Random-Liu Do you know how well crazy variable names are supported on the CRI API and nodes? (I have no idea what I'm talking about, I'm repeating words Jordan used).

The change should be okay as far as CRI is concerned. As for kubelet, it uses the same validation method `IsEnvVarName`, so it should be okay.
To be safe, you can change one of the env variable e2e tests to make sure this works properly.

/cc @PatrickLang @feiskyer to make sure this works on Windows too.

lavalamp commented:
Doesn't that mean apiserver might accept a pod that an old kubelet will
reject?

On Wed, Oct 17, 2018, 3:31 PM Yu-Ju Hong <notifications@github.com> wrote:

> @yujuhong <https://github.com/yujuhong> @Random-Liu
> <https://github.com/Random-Liu> Do you know how well crazy variable names
> are supported on the CRI API and nodes? (I have no idea what I'm talking
> about, I'm repeating words Jordan used).
>
> The change should be okay as far as CRI is concerned. As for kubelet, it
> uses the same validation method IsEnvVarName, so it should be okay.
> To be safe, you can change one of the env variable e2e tests to make sure
> this works properly.
>
> /cc @PatrickLang <https://github.com/PatrickLang> @feiskyer
> <https://github.com/feiskyer> to make sure this works on Windows too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/69415#issuecomment-430811826>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngluxyQfqr46muQSw9ewDfPPqRl7Olks5ul6_UgaJpZM4XHqxe>
> .
>

liggitt commented:
> As for kubelet, it uses the same validation method IsEnvVarName, so it should be okay.

That means version skewed kubelets will fail validation. Does that happen gracefully or does the kubelet lose its mind?

yujuhong commented:
> That means version skewed kubelets will fail validation. Does that happen gracefully or does the kubelet lose its mind?

Yep. It will fail validation.

Kubelet won't lose its mind, but it will skip the invalid env variables, and also send an event to warn the users.

lavalamp commented:
That's not catastrophic but it's not great, either. At a minimum it needs a big warning in the release note.

apelisse commented:
So I guess we need a bigger warning in the release note, and have an example in an e2e tests to make sure that everything works nicely. 

And I think we should relax it even further because we don't want to break compatibility every time we want to add a new character?

yujuhong commented:
> But will still run the pod? Can you point to where that happens? I know kubelet used to have more severe issues with pods that didn't validate.

Where kubelet validates the env variables:
  https://github.com/kubernetes/kubernetes/blob/feb6475a3097cf6c7b7750c7899eb13ffca17d37/pkg/kubelet/kubelet_pods.go#L542

kubelet also calls `ValidatePod()`, but only for non-apiserver pods. Kubelet's behavior should be easy to verify manually, and that'd probably be more reliable than my memory/code search :-)

PatrickLang commented:
Seems ok from a Windows standpoint - only `=` is documented as prohibited - [source](https://docs.microsoft.com/en-us/windows/desktop/ProcThread/environment-variables). 

I have never seen it in use though, and I don't see it as a need. Most Windows processes use _ as a delimiter instead of : in environment variable names.

vithati commented:
> So I guess we need a bigger warning in the release note, and have an example in an e2e tests to make sure that everything works nicely.
> 
> And I think we should relax it even further because we don't want to break compatibility every time we want to add a new character?

I have relaxed validation condition to accept all character except starting with a digit. I have modified existing e2e conformance test to validate new characters in variable names and updated the release notes.

bgrant0607 commented:
Please see discussion in https://github.com/kubernetes/kubernetes/issues/53201

bgrant0607 commented:
cc @kow3ns

thockin commented:
#53201 wrongly asserts that `:` is not POSIX compatible.  It is.

bgrant0607 commented:
@thockin If we're going to change this, we need someone from SIG Node to think hard about what might break and what additional tests we may need.

From an API perspective, changing validation in any way always has the potential to break clients and normally would be forbidden, so we'll need to weigh that risk against fixing this bug:
https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#backward-compatibility-gotchas

cc @liggitt on API validation change
cc @yujuhong on Node impact

bgrant0607 commented:
/area conformance

apelisse commented:
@vithati Thank you so much for working on this, it looks like `=` is not allowed [1]
> These strings have the form name=value; names shall not contain the character '='.

@yujuhong Can you think of other tests we would have to change/add to test this? The PR currently has `test/e2e/common/{configmap.go,secret.go,expansion.go}`

[1]: http://pubs.opengroup.org/onlinepubs/009604599/basedefs/xbd_chap08.html

vithati committed:
Except '=' character allow all characters in environment variable name

vithati committed:
Added test for invalid prefix key name.

vithati commented:
I have modified envVarNameFmt  regex to except all characters except '='. Let me know if i need to add/modify more test cases to cover the scenarios or any other validation needed to validate the fix.

vithati commented:
> @vithati Thank you so much for working on this, it looks like `=` is not allowed [1](http://pubs.opengroup.org/onlinepubs/009604599/basedefs/xbd_chap08.html)
> 
> > These strings have the form name=value; names shall not contain the character '='.
> 
> @yujuhong Can you think of other tests we would have to change/add to test this? The PR currently has `test/e2e/common/{configmap.go,secret.go,expansion.go}`

I have modified envVarNameFmt regex to except all characters except '=' in environment variable name. Let me know if i need to add/modify more test cases to cover the scenarios or any other validation needed to validate the fix.

bgrant0607 commented:
/hold

smarterclayton commented:
Going from a strict set to “everything” is scary because we have no way to take that back once it’s part of the public API.  I’d rather continue to selectively add things.

jbeda commented:
To echo what @smarterclayton is saying -- it feels to me like this is an easy place for us to introduce security vulnerabilities depending on how we set or pass these things down.  I don't have anything concrete in mind but [env variables are scary from a security PoV](https://www.cvedetails.com/google-search-results.php?q=sudo+environment).

thockin commented:
Those security concerns are about how env vars are used and not about the shape/form of them.

This PR is exclusively about the shape/form.

I don't think it is useful of friendly to force people to PR each new character we want to support.   Every time we have this incompatible-potential event.  We'd do better to follow POSIX.  

Now, that said, the RIGHT answer is to do OS-dependent validation, but we don't know the target OS at this point, so maybe we should validate the intersection of OSes?  But that means adding a new OS could restrict the allowed-set, which is verboten.  That seems unlikely.

Kubelet ultimately has to defend against bad pods, so it should probably be Kubelet that has final say in validating and rejecting (or omitting) bad env vars.  At that point, the only character that seems universally disallowed is `=`.  POSIX has a list of compatible characters (all ASCII), which includes control characters.  If we feel the need to limit it to printable characters, I could live with that.

E.g. `unicode.IsPrint(r) && r < unicode.MaxASCII`.  We can debate `IsPrint()` vs `IsGraphic()`, for which the delta seems to be tabs. 

That makes it not "everything" but a broader starting point.

thockin commented:
@vithati thoughts on my last comment?

vithati commented:
> @vithati thoughts on my last comment?
 I agree with you, as you mentioned we can start with IsPrint() (or IsGraphic()) instead of broader set.
JFI, I see container runtimes like lxc, runc not have restrictions on env variable names, but appc has same restriction as Kubernetes currently.

thockin commented:
What is appc - rkt?  I don't know that this matters.

If this PR was retooled to allow all printable ASCII except '=', including the apiserver and kubelet validation, I would be inclined to approve it.  We should probably add it under a feature gate so if it DOES cause problems, we can flip back to old mode quickly.  Not sure if we even need to start with Alpha, maybe just starting at Beta is OK, but we can discuss that.

apelisse commented:
> We should probably add it under a feature gate so if it DOES cause problems, we can flip back to old mode quickly.

Wouldn't this make it worse from a backward/forward compatibility perspective?

thockin commented:
I meant on a per-user basis.  It is delicate in that we don't want to
break anyone with stored state - maybe we should spend more time
thinking about how to gain confidence in this change?  I don't see a
problem with it, but I guess that's the definition of surprise :)

On Mon, Jan 7, 2019 at 2:07 PM Antoine Pelisse <notifications@github.com> wrote:
>
> We should probably add it under a feature gate so if it DOES cause problems, we can flip back to old mode quickly.
>
> Wouldn't this make it worse from a backward/forward compatibility perspective?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or mute the thread.

liggitt commented:
> If this PR was retooled to allow all printable ASCII except '=', including the apiserver and kubelet validation, I would be inclined to approve it. We should probably add it under a feature gate so if it DOES cause problems, we can flip back to old mode quickly. Not sure if we even need to start with Alpha, maybe just starting at Beta is OK, but we can discuss that.

We'd need to make sure we tolerated any new characters in update validation for a release prior to allowing them on create, to protect against a situation where a skewed HA cluster could have half the apiservers consider objects with the new characters persisted in etcd to be invalid.

thockin commented:
As much as it pains me to turn a trivial PR into a 6-month effort, I agree
in principle.  We really have 2 options for such a change.

1) Release 1.X which accepts and uses the extended character set, but
doesn't allow it through API validation.  Release 1.Y which allows it.  Any
cluster rollback from 1.Y to 1.X would break API calls, but not break
running clusters.  HA master upgrades would be OK.  We have no real
precedent for this, but it has been discussed in sig-arch.

2) Release 1.X which accepts and uses the extended character set and which
has a flag-gate controlling API validation, disabled by default.  Once HA
upgrade is done, flip the flag.  Flipping the flag back would break API
calls, but not break running clusters.  We have no real precedent for this
pattern either, AFAIK.

This seems like such an innocuous change to be the victim of evolving
policy, but we have to start somewhere.

Thoughts?

On Mon, Jan 7, 2019 at 8:34 PM Jordan Liggitt <notifications@github.com>
wrote:

> If this PR was retooled to allow all printable ASCII except '=', including
> the apiserver and kubelet validation, I would be inclined to approve it. We
> should probably add it under a feature gate so if it DOES cause problems,
> we can flip back to old mode quickly. Not sure if we even need to start
> with Alpha, maybe just starting at Beta is OK, but we can discuss that.
>
> We'd need to make sure we tolerated any new characters in update
> validation for a release prior to allowing them on create, to protect
> against a situation where a skewed HA cluster could have half the
> apiservers consider objects with the new characters persisted in etcd to be
> invalid.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/69415#issuecomment-452172161>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVIskbB7jP2Ta2T11FtmaSFu7PYgKks5vBB_agaJpZM4XHqxe>
> .
>

bgrant0607 commented:
@thockin What do you mean by "Release 1.X which accepts and uses the extended character set, but
doesn't allow it through API validation"? This PR just changes validation, documentation, and tests. If validation were not changed, I'm not sure what "accepts and uses" would mean.

bgrant0607 commented:
As for enabling after HA upgrade, we're going to need to do that more generally:
https://github.com/kubernetes/kubernetes/issues/4855#issuecomment-360928086

We should also consider where kubelets apply validation. Just to static pods?

thockin commented:
@bgrant0607 I meant that 1.x release would tolerate and use any extended values it finds in etcd, but would not allow use of extended values in API mutations.  I.e. if I upgrade to 1.y and rollback to 1.x or if I have version skew it will be OK.

Agree re: kubelet

liggitt commented:
I'd envision a rollout like this:

In version X:
1. handle the expanded values, and tolerate persisted objects with the expanded values:
  1a: update validation to validate the expanded values
  1b: update consumers of the field to handle the expanded values
  1c: allow an update of an object already using the expanded values to continue using them
2. prevent new use of the expanded values:
  2a: prevent creation of a new object using the expanded values
  2b: prevent an update of an object not yet using the expanded values from starting to use them

In some later version, the restrictions on create and update (2a,2b) can be removed.
* This should be done no earlier than X+1 (so a skewed HA control plane does not consider persisted data invalid, and single version control plane rollback is safe).
* This could require a longer waiting period, depending on the consumers that were updated in 1a, what behavior they had prior to being updated, and what level of skew is supported.
  * Did they skip the new data in a safe way? (e.g. not set the new envvar)
  * Did they skip the new data in an unsafe way? (e.g. format a volume as a filesystem instead of a block device... if so, this is probably a sign the API change is not a safe one to make at all)
  * Did they hard fail? Could require waiting until version X when the new data was handled is the oldest supported version in the cluster.

This is the same process we have to follow when allowing a new enum value (e.g. adding "SCTP" as a valid port protocol).

In this case, it appears kubelets do not validate envvar names coming from the API, and only warn on invalid envvar names resulting from secret/configmap -> envvar mappings. That means we could relax validation in 1.14, and remove create/update restrictions in 1.15.

thockin commented:
@vithati any appetite for trying to get this into 1.14 ?

vithati commented:
> @vithati any appetite for trying to get this into 1.14 ?

@thockin, I can't make it for 1.14. I will work on this for 1.15.

alex-slynko commented:
Hi @vithati 

Is there any news about the PR?
We tried to use the variable with `:` to configure npm registry for building image, but it failed. We have implemented a workaround, but it would be nice to have a proper fix.

spiffxp commented:
/priority awaiting-more-evidence

vithati commented:
I am not able to get time to work on the changes. Can someone take this up?

spiffxp commented:
/remove-area conformance

hpandeycodeit commented:
@vithati I would like to resume working on this PR. Let me know if that's okay with you.

vithati commented:
> @vithati I would like to resume working on this PR. Let me know if that's okay with you.



> @vithati I would like to resume working on this PR. Let me know if that's okay with you.

Thanks, you can take it up.

liggitt commented:
/close

in favor of https://github.com/kubernetes/kubernetes/pull/83032
